### PR TITLE
[SPARK-33599][SQL] Group exception messages in catalyst/analysis

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR group exception messages in `/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis`.

### Why are the changes needed?
It will largely help with standardization of error messages and its maintenance.

### Does this PR introduce _any_ user-facing change?
No. Error messages remain unchanged.

### How was this patch tested?
No new tests - pass all original tests to make sure it doesn't break any existing behavior.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -326,4 +340,170 @@ object QueryCompilationErrors {
...
+    new AnalysisException(s"$hintName Hint expects a partition number as a parameter")
+  }
+
+  def attributeNameSyntaxError(name: String): Throwable = {
+    new AnalysisException(s"syntax error in attribute name: $name")
+  }
+
+  def onlyStarExpandStructDataTypesError(target: Option[Seq[String]]): Throwable = {
+    new AnalysisException("Can only star expand struct data types. Attribute: `" +
+      target.get + "`")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:413]: nit `new AnalysisException(s"Can only star expand struct data types. Attribute: `${target.get}`")`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -326,4 +340,170 @@ object QueryCompilationErrors {
...
+
+  def cannotDropViewWithDropTableError(): Throwable = {
+    new AnalysisException("Cannot drop a view with DROP TABLE. Please use DROP VIEW instead")
+  }
+
+  def showColumnsWithConflictDatabasesError(
+      db: Seq[String], v1TableName: TableIdentifier): Throwable = {
+    new AnalysisException(
+      "SHOW COLUMNS with conflicting databases: " +
+        s"'${db.head}' != '${v1TableName.database.get}'")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:475]: nit:
```
    new AnalysisException("SHOW COLUMNS with conflicting databases: " +
        s"'${db.head}' != '${v1TableName.database.get}'")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  def cannotGenerateCodeForExpressionError(generator: UnresolvedGenerator): Throwable = {
+    new UnsupportedOperationException(s"Cannot generate code for expression: $generator")
+  }
+
+  def cannotTerminateExpressionError(generator: UnresolvedGenerator): Throwable = {
+    new UnsupportedOperationException(s"Cannot terminate expression: $generator")
+  }
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:59]: nit: remove this blank.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 1 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:413]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:475]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala]: OK

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
cc @allisonwang-db @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 4]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+
+  def replaceColumnsOnlySupportedWithV2TablesError(): Throwable = {
+    new AnalysisException("REPLACE COLUMNS is only supported with v2 tables.")
+  }
+
+  def alterQualifiedColumnOnlySupportedWithV2TablesError(): Throwable = {
+    new AnalysisException("ALTER COLUMN with qualified column is only supported with v2 tables.")
+  }
+
+  def alterColumnWithV1TablesCannotSpecifyNotNullError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:432]: ditto (drop `s` in `tables`)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 5]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException("ALTER COLUMN with v1 tables cannot specify NOT NULL.")
+  }
+
+  def alterOnlySupportedWithV2TablesError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:436]: ditto (drop `s` in `tables`)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 6]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException("ALTER COLUMN ... FIRST | ALTER is only supported with v2 tables.")
+  }
+
+  def alterColumnCannotFindColumnInV1TableError(colName: String, v1Table: V1Table): Throwable = {
+    new AnalysisException(
+      s"ALTER COLUMN cannot find column $colName in v1 table. " +
+        s"Available: ${v1Table.schema.fieldNames.mkString(", ")}")
+  }
+
+  def renameColumnOnlySupportedWithV2TablesError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:446]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 7]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException("RENAME COLUMN is only supported with v2 tables.")
+  }
+
+  def dropColumnOnlySupportedWithV2TablesError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:450]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 8]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException("DROP COLUMN is only supported with v2 tables.")
+  }
+
+  def invalidDatabaseNameError(quoted: String): Throwable = {
+    new AnalysisException(s"The database name is not valid: $quoted")
+  }
+
+  def replaceTableOnlySupportedWithV2TablesError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:458]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 9]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException("REPLACE TABLE is only supported with v2 tables.")
+  }
+
+  def replaceTableAsSelectOnlySupportedWithV2TablesError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:462]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    new IllegalStateException(
+      "Internal error: logical hint operator should have been removed during analysis")
+  }
+
+  def logicalPlanHaveOutputOfCharOrVarcharError(leaf: LeafNode): Throwable = {
+    new IllegalStateException(
+      s"[BUG] logical plan should not have output of char/varchar type: $leaf")
+  }
+
+  def cannotEvaluateExpressionError(generator: UnresolvedGenerator): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:48]: `cannotEvaluateGeneratorError` is probably more specific

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    new UnsupportedOperationException(s"Cannot evaluate expression: $generator")
+  }
+
+  def cannotGenerateCodeForExpressionError(generator: UnresolvedGenerator): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:52]: cannotGenerateCodeForGeneratorError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    new UnsupportedOperationException(s"Cannot generate code for expression: $generator")
+  }
+
+  def cannotTerminateExpressionError(generator: UnresolvedGenerator): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:56]: cannotTerminateGeneratorError

**Review by allisonwang-db — COMMENTED** [review]
Looks good! Left a few comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 13]
```diff
@@ -196,6 +200,11 @@ object QueryCompilationErrors {
       t.origin.line, t.origin.startPosition)
   }
 
+  def nonStreamingTempViewError(quoted: String): Throwable = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:203]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 14]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+        s"${queryColumnNames.mkString("[", ",", "]")}")
+  }
+
+  def attributeNotFoundError(colName: String, child: LogicalPlan): Throwable = {
+    new AnalysisException(
+      s"Attribute with name '$colName' is not found in " +
+        s"'${child.output.map(_.name).mkString("(", ",", ")")}'")
+  }
+
+  def cannotUpCastAsAttributeTruncateError(
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:357]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 15]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+      originAttr: Attribute, catalogString: String): Throwable = {
+    new AnalysisException(s"Cannot up cast ${originAttr.sql} from " +
+      s"${originAttr.dataType.catalogString} to $catalogString as it may truncate\n")
+  }
+
+  def functionUndefinedError(name: FunctionIdentifier): Throwable = {
+    new AnalysisException(s"undefined function $name")
+  }
+
+  def invalidFunctionArgumentsError(
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:367]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 16]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+
+  def functionAcceptsOnlyOneArgumentError(name: String): Throwable = {
+    new AnalysisException(s"Function $name accepts only one argument")
+  }
+
+  def alterV2TableSetLocationWithPartitionNotSupportedError(): Throwable = {
+    new AnalysisException("ALTER TABLE SET LOCATION does not support partition for v2 tables.")
+  }
+
+  def joinStrategyHintParameterNotExpectedError(unsupported: Any): Throwable = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:392]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 17]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException("Join strategy hint parameter " +
+      s"should be an identifier or string but was $unsupported (${unsupported.getClass}")
+  }
+
+  def hintParameterNotIncludeColumnsError(
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:397]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 18]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+      hintName: String, invalidParams: Seq[Any]): Throwable = {
+    new AnalysisException(s"$hintName Hint parameter should include columns, but " +
+      s"${invalidParams.mkString(", ")} found")
+  }
+
+  def hintParameterNotExpectedError(hintName: String): Throwable = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:403]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 19]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException(s"$hintName Hint expects a partition number as a parameter")
+  }
+
+  def attributeNameSyntaxError(name: String): Throwable = {
+    new AnalysisException(s"syntax error in attribute name: $name")
+  }
+
+  def onlyStarExpandStructDataTypesError(target: Option[Seq[String]]): Throwable = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:411]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 20]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException(s"Can only star expand struct data types. Attribute: `${target.get}`")
+  }
+
+  def cannotResolveTargetAccordingGivenInputColumnsError(
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:415]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 21]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+      targetString: String, from: String): Throwable = {
+    new AnalysisException(s"cannot resolve '$targetString.*' given input columns '$from'")
+  }
+
+  def addColumnWithV1TableCannotSpecifyNotNullError(): Throwable = {
+    new AnalysisException("ADD COLUMN with v1 tables cannot specify NOT NULL.")
+  }
+
+  def replaceColumnsOnlySupportedWithV2TablesError(): Throwable = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:424]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 22]
```diff
@@ -326,4 +340,164 @@ object QueryCompilationErrors {
...
+    new AnalysisException("REPLACE COLUMNS is only supported with v2 tables.")
+  }
+
+  def alterQualifiedColumnOnlySupportedWithV2TablesError(): Throwable = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:428]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 4 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:432]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 5 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:436]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 6 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:446]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 7 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:450]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 8 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:458]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 9 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:462]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala (hunk 10 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:48]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala (hunk 11 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:52]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala (hunk 12 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:56]: OK

**Comment by beliefer** [comment]
@allisonwang-db cc @cloud-fan @HyukjinKwon @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (allisonwang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:357]: We can also remove "Truncate" in the method name here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala [hunk 23]
```diff
@@ -326,4 +340,165 @@ object QueryCompilationErrors {
...
+    new AnalysisException(
+      s"Attribute with name '$colName' is not found in " +
+        s"'${child.output.map(_.name).mkString("(", ",", ")")}'")
+  }
+
+  def cannotUpCastAsAttributeTruncateError(
+      fromAttr: Attribute, toAttr: Attribute): Throwable = {
+    new AnalysisException(s"Cannot up cast ${fromAttr.sql} from " +
+      s"${fromAttr.dataType.catalogString} to ${toAttr.dataType.catalogString} " +
+      "as it may truncate\n")
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:361]: nit: no need `\n` in the end.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+ * Currently it includes all [[SparkException]]s and RuntimeExceptions(e.g.
+ * UnsupportedOperationException, IllegalStateException).
+ */
+object QueryExecutionErrors {
+
+  def columnChangeUnsupportedError(): Throwable = {
+    new UnsupportedOperationException("Please add an implementation for a column change here")
+  }
+
+  def unexpectedPlanReturnError(other: LogicalPlan, methodName: String): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:34]: nit: `other: LogicalPlan` -> `plan: LogicalPlan`

**Review by allisonwang-db — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala (hunk 23 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala:361]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala (hunk 24 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/QueryExecutionErrors.scala:34]: OK

**Comment by beliefer** [comment]
cc @cloud-fan @allisonwang-db

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan and @beliefer . Could you check the GitHub Action?

**Comment by cloud-fan** [comment]
hmm, I thought the GA only brings extra Scala 2.13 check than jenkins. @beliefer can you check the docgen error in GitHub Action?

**Comment by cloud-fan** [comment]
@dongjoon-hyun seems GA for the master branch is healthy: https://github.com/apache/spark/commits/master . This was probably transient errors.

**Pull request closed without merging**